pkg/compactor/block_upload.go

```diff
@@ -559,3 +697,53 @@ func (c *MultitenantCompactor) loadValidation(ctx context.Context, userBkt objst
 
 	return v, nil
 }
+
+func (c *MultitenantCompactor) uploadValidationWithError(ctx context.Context, blockID ulid.ULID,
+	userBkt objstore.Bucket, errorStr string) error {
+	val := validationFile{
+		LastUpdate: time.Now().UnixMilli(),
+		Error:      errorStr,
+	}
+	dst := path.Join(blockID.String(), validationFilename)
+	if err := marshalAndUploadToBucket(ctx, userBkt, dst, val); err != nil {
+		return errors.Wrapf(err, "failed uploading %s to bucket", validationFilename)
+	}
+	return nil
+}
+
+func (c *MultitenantCompactor) uploadValidation(ctx context.Context, logger log.Logger, blockID ulid.ULID,
+	userBkt objstore.Bucket) error {
+	return c.uploadValidationWithError(ctx, blockID, userBkt, "")
+}
+
+func (c *MultitenantCompactor) periodicValidationUpdater(ctx context.Context, logger log.Logger, blockID ulid.ULID,
+	userBkt objstore.Bucket, ch chan struct{}, wg *sync.WaitGroup, cancelFn func()) {
+	defer wg.Done()
+	ticker := time.NewTicker(validationHeartbeatInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ch:
+			return
+		case <-ctx.Done():
+			return
```

Comment:
Instead of having two ways of stopping this method (channel + context), we could only use context.

Reply:
Agreed, I had added the context and this seemed strange. I removed the channel.